src/Conversion/KrnlToLLVM/KrnlEntryPoint.cpp

```diff
@@ -426,13 +458,212 @@ class KrnlEntryPointOpLowering : public OpRewritePattern<KrnlEntryPointOp> {
     }
     return SymbolRefAttr::get(ctx, funcName);
   }
+
+  // Emit code for `IF lhs != rhs THEN return null ELSE do nothing`
+  void equalOrFailed(PatternRewriter &rewriter, Location loc, Value lhs,
+      Value rhs, std::string errorMsg = "", bool appendRHS = true) const {
+    // Split the current block into IF, THEN, END blocks.
+    Block *ifBlock, *thenBlock, *endBlock;
+    ifBlock = rewriter.getInsertionBlock();
+    thenBlock = ifBlock->splitBlock(rewriter.getInsertionPoint());
+    endBlock = rewriter.createBlock(
+        thenBlock->getParent(), std::next(Region::iterator(thenBlock)));
+
+    // Emit code for the IF block.
+    rewriter.setInsertionPointToEnd(ifBlock);
+    Value failed =
+        rewriter.create<LLVM::ICmpOp>(loc, LLVM::ICmpPredicate::ne, lhs, rhs);
+
+    // Branch the block into the THEN and END blocks.
+    rewriter.create<LLVM::CondBrOp>(
+        loc, failed, thenBlock, ValueRange(), endBlock, ValueRange());
+
+    // Emit code for the THEN block: return NULL.
+    rewriter.setInsertionPointToStart(thenBlock);
+    KrnlBuilder createKrnl(rewriter, loc);
+    if (appendRHS)
+      createKrnl.printf(StringRef(errorMsg), rhs, rewriter.getI64Type(), true);
+    else
+      createKrnl.printf(StringRef(errorMsg + "\n"));
+    Value nullPtr = rewriter.create<LLVM::NullOp>(
+        loc, LLVM::LLVMPointerType::get(rewriter.getI8Type()));
+    rewriter.create<LLVM::ReturnOp>(loc, ArrayRef<Value>({nullPtr}));
+
+    // Emit code for the END block: continue with other generated code.
+    rewriter.setInsertionPointToStart(endBlock);
+  }
+
+  void emitVerificationCodeForInputTensors(PatternRewriter &rewriter,
+      Location loc, const RuntimeAPIRegistry &apiRegistry, Value wrappedInput,
+      StringRef inSigJSON) const {
+    Type int64Ty = rewriter.getI64Type();
+    Type opaquePtrTy = LLVM::LLVMPointerType::get(rewriter.getI8Type());
+
+    auto JSONInput = llvm::json::parse(inSigJSON.data());
```

Comment:
Just requesting a clarification: here we don't have any JSON calls at runtime, correct?  The compiler parse the JSON string at compile time to generate a test that will compare the string from the signature with the expected string found at compile time from JSON?
So it feels like you implemented the efficient test, isn't it?

Reply:
Correct. We parse the JSON string at compile time to get the element type, rank and shape, and then generate runtime code to compare them with the ones from users. It is integer comparison, so quite efficient at runtime.

We can avoid parsing the JSON string at compile time also by storing information about input MemRefs before they are unfolded into LLVM struct. But, I feel that it's unncessary, given that the JSON string correctly reflects input signature.